From 1da8826b860007f687b4dc54e894a333d9519925 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Wed, 3 Jul 2024 11:47:16 +0200
Subject: [PATCH 1011/1695] drivers: stm32_remoteproc: use
 firewall_dt_get_alternate_conf API

Utilize the `firewall_dt_get_alternate_conf` function to store the two
alternative configurations for the memory region:
- "default": Configuration applied when the remote processor is running.
  The code and data memory regions should be isolated for Cortex-M usage.
- "load": Configuration used to authenticate and load the firmware.
  This configuration must ensure that Cortex-A has secure and exclusive
  access.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 98b23392a55ca18ec317010767ce721d3f758b98 drivers: stm32_remoteproc: use firewall_dt_get_alternate_conf API for stm32mp15
2517b951d8a66398a0ad92fa3d056ce7c8ceb1c6 drivers: stm32_remoteproc: use firewall_dt_get_alternate_conf API]
Change-Id: Ieea21b70761ed8debb88f3820b2c7557f5956fb9
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391543
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/remoteproc/stm32_remoteproc.c | 60 +++++++++++++++++-----
 1 file changed, 47 insertions(+), 13 deletions(-)

diff --git a/core/drivers/remoteproc/stm32_remoteproc.c b/core/drivers/remoteproc/stm32_remoteproc.c
index 46a9f8e57..c5257dd91 100644
--- a/core/drivers/remoteproc/stm32_remoteproc.c
+++ b/core/drivers/remoteproc/stm32_remoteproc.c
@@ -5,6 +5,7 @@
 
 #include <assert.h>
 #include <config.h>
+#include <drivers/firewall_device.h>
 #include <drivers/rstctrl.h>
 #include <drivers/stm32_remoteproc.h>
 #include <keep.h>
@@ -23,6 +24,9 @@
 
 #define INITVTOR_MASK	GENMASK_32(31, 7)
 
+#define FIREWALL_CONF_DEFAULT "default"
+#define FIREWALL_CONF_LOAD    "load"
+
 /**
  * struct stm32_rproc_mem - Memory regions used by the remote processor
  *
@@ -30,11 +34,16 @@
  * @da:		device address corresponding to the physical base address
  *		from remote processor space perspective
  * @size:	size of the region
+ * @default_conf: default firewall configuration applied on memory
+ * @load_conf:	  firewall configuration applied on memory to give access to the
+ *		  TEE remoteproc framework to load the firmware
  */
 struct stm32_rproc_mem {
 	paddr_t addr;
 	paddr_t da;
 	size_t size;
+	struct firewall_alt_conf *default_conf;
+	struct firewall_alt_conf *load_conf;
 };
 
 /**
@@ -401,6 +410,26 @@ TEE_Result stm32_rproc_enable_sec_boot(uint32_t rproc_id)
 	return TEE_SUCCESS;
 }
 
+static void stm32_rproc_free_regions(struct stm32_rproc_instance *rproc)
+{
+	struct stm32_rproc_mem *regions = rproc->regions;
+	int i = 0;
+
+	if (!regions)
+		return;
+
+	for (i = 0; i < (int)rproc->n_regions; i++) {
+		if (regions[i].default_conf)
+			firewall_alternate_conf_put(regions[i].default_conf);
+		if (regions[i].load_conf)
+			firewall_alternate_conf_put(regions[i].load_conf);
+	}
+
+	free(regions);
+	rproc->n_regions = 0;
+	rproc->regions = NULL;
+}
+
 /* Get device tree memory regions reserved for the Cortex-M and the IPC */
 static TEE_Result stm32_rproc_parse_mems(struct stm32_rproc_instance *rproc,
 					 const void *fdt, int node)
@@ -431,13 +460,16 @@ static TEE_Result stm32_rproc_parse_mems(struct stm32_rproc_instance *rproc,
 	if (!regions)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
+	rproc->n_regions = n_regions;
+	rproc->regions = regions;
+
 	for (i = 0; i < n_regions; i++) {
 		int pnode = 0;
 
 		pnode = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(list[i]));
 		if (pnode < 0) {
 			res = TEE_ERROR_GENERIC;
-			goto err;
+			return res;
 		}
 
 		regions[i].addr = fdt_reg_base_address(fdt, pnode);
@@ -445,36 +477,38 @@ static TEE_Result stm32_rproc_parse_mems(struct stm32_rproc_instance *rproc,
 
 		if (regions[i].addr <= 0 || regions[i].size <= 0) {
 			res = TEE_ERROR_GENERIC;
-			goto err;
+			return res;
 		}
+		res = firewall_dt_get_alternate_conf(fdt, pnode,
+						     FIREWALL_CONF_DEFAULT,
+						     &regions[i].default_conf);
+		if (res)
+			return res;
+		res = firewall_dt_get_alternate_conf(fdt, pnode,
+						     FIREWALL_CONF_LOAD,
+						     &regions[i].load_conf);
+		if (res)
+			return res;
 
 		res = stm32_rproc_get_dma_range(&regions[i], fdt, node);
 		if (res)
-			goto err;
+			return res;
 
 		if (!regions[i].addr || !regions[i].size) {
 			res = TEE_ERROR_BAD_PARAMETERS;
-			goto err;
+			return res;
 		}
 
 		DMSG("register region %#"PRIxPA" size %#zx",
 		     regions[i].addr, regions[i].size);
 	}
 
-	rproc->n_regions = n_regions;
-	rproc->regions = regions;
-
 	return TEE_SUCCESS;
-
-err:
-	free(regions);
-
-	return res;
 }
 
 static void stm32_rproc_cleanup(struct stm32_rproc_instance *rproc)
 {
-	free(rproc->regions);
+	stm32_rproc_free_regions(rproc);
 	free(rproc);
 }
 
-- 
2.47.2

